bluetooth/src/commonMain/kotlin/scanner/ScanningState.kt

```diff
@@ -267,30 +328,120 @@ sealed interface ScanningState : KalugaState {
      * A [ScanningState] indicating the system does not support Bluetooth
      */
     interface NoHardware : ScanningState
+
+    val nothingFound: Devices
 }
 
-internal sealed class ScanningStateImpl {
+/**
+ * Transitions into a [Scanning] state where a [Device] is added or updated
+ * @param identifier the [Identifier] of the [Device] discovered
+ * @param rssi the [RSSI] value of the [Device] discovered
+ * @param advertisementData the [BaseAdvertisementData] of the [Device] discovered
+ * @param deviceCreator Method for creating a [Device] if it had not been scanned previously.
+ * @return method for transitioning into a [Scanning] state where the [Device] is discovered
+ */
+suspend fun Scanning.discoverDevice(
+    identifier: Identifier,
+    rssi: RSSI,
+    advertisementData: BaseAdvertisementData,
+    deviceCreator: () -> Device
+) = discoverDevices(listOf(Scanning.DiscoveredDevice(identifier, rssi, advertisementData, deviceCreator)))
+
+/**
+ * Default implementation of [ScanningState.Devices]
+ */
+data class DefaultDevices(
+    override val allDevices: Map<Identifier, Device>,
+    override val identifiersFoundForDeviceDiscoveryMode: Map<ScanningState.DeviceDiscoveryMode, Set<Identifier>>,
+    override val currentScanFilter: ScanningState.DeviceDiscoveryMode.Scanning
+) : ScanningState.Devices {
+
+    /**
+     * Constructor
+     * @param filter the [Filter] to use for the [currentScanFilter]
+     */
+    constructor(filter: Filter) : this(emptyMap(), emptyMap(), ScanningState.DeviceDiscoveryMode.Scanning(filter))
+
+    override fun copyAndAddScanned(
+        identifier: Identifier,
+        createDevice: () -> Device
+    ): ScanningState.Devices = copyAndAdd(identifier, currentScanFilter, createDevice)
+
+    override fun copyAndSetPaired(
+        devices: Map<Identifier, () -> Device>,
+        filter: Filter,
+        removeForAllPairedFilters: Boolean
+    ): ScanningState.Devices {
+        val filtersToRemove = if (removeForAllPairedFilters) {
+            identifiersFoundForDeviceDiscoveryMode.keys.filterIsInstance<ScanningState.DeviceDiscoveryMode.Paired>()
+        } else {
+            listOf(ScanningState.DeviceDiscoveryMode.Paired(filter))
+        }
 
-    companion object {
-        val nothingFound = Devices(filter = emptySet())
+        val newPairedDevices = devices.map { (identifier, createDevice) -> identifier to allDevices.getOrElse(identifier, createDevice) }
+        val deviceWithFiltersRemoved = filtersToRemove.fold(this) { acc, paired ->
+            acc.cleanFilter(paired)
+        }
+        val newIdentifiersForDiscoveryMode = deviceWithFiltersRemoved.identifiersFoundForDeviceDiscoveryMode.toMutableMap().apply {
+            put(ScanningState.DeviceDiscoveryMode.Paired(filter), devices.keys)
+        }.toMap()
+
+        val newDevices = deviceWithFiltersRemoved.allDevices.toMutableMap().apply {
+            putAll(newPairedDevices.toMap())
+        }
```

Comment:
This remains mutable map.
However you can simplify (and above as well)
```
val newDevices = deviceWithFiltersRemoved.allDevices + newPairedDevices.toMap()
```

Plus operator should work on immutable maps like put or putAll